finality-aleph/src/sync/mod.rs

```diff
@@ -0,0 +1,92 @@
+use std::{
+    fmt::{Debug, Display},
+    hash::Hash,
+};
+
+mod substrate;
+
+/// The identifier of a block, the least amount of knowledge we can have about a block.
+pub trait BlockIdentifier: Clone + Hash + Debug + Eq {
+    /// The block number, useful when reasoning about hopeless forks.
+    fn number(&self) -> u32;
+}
+
+/// Informs the sync that it should attempt to acquire the specified data.
+pub trait Requester<BI: BlockIdentifier> {
+    /// The sync should attempt to acquire justifications for this block.
+    fn request_justification(&self, id: BI);
+}
+
+/// The header of a block, containing information about the parent relation.
+pub trait Header: Clone {
+    type Identifier: BlockIdentifier;
+
+    /// The identifier of this block.
+    fn id(&self) -> Self::Identifier;
+
+    /// The identifier of this block's parent.
+    fn parent_id(&self) -> Option<Self::Identifier>;
+}
+
+/// The justification of a block, including a header.
+pub trait Justification: Clone {
```

Comment:
suggestion: Add method returning raw_justification here and in Verifier just pass Justification instead of header + raw_justification

Reply:
Hm, my idea was to have this work as a constructor, but your comment made me realize that this is incoherent with sending justifications with headers over the network...

Changed it to maybe be better now – `Justification` has an associated type for an unverified version, can be changed into that and verifier should be the only way of performing the change the other way. Amusingly this should give us some type safety even if we set `Unverified = Self` in the implementation, so that's nice.